##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# 케라스와 텐서플로 허브를 사용한 영화 리뷰 텍스트 분류하기

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />TensorFlow.org에서 보기</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/ko/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />구글 코랩(Colab)에서 실행하기</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs-l10n/blob/master/site/ko/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />깃허브(GitHub) 소스 보기</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs-l10n/site/ko/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>

Note: 이 문서는 텐서플로 커뮤니티에서 번역했습니다. 커뮤니티 번역 활동의 특성상 정확한 번역과 최신 내용을 반영하기 위해 노력함에도
불구하고 [공식 영문 문서](https://www.tensorflow.org/?hl=en)의 내용과 일치하지 않을 수 있습니다.
이 번역에 개선할 부분이 있다면
[tensorflow/docs-l10n](https://github.com/tensorflow/docs-l10n/) 깃헙 저장소로 풀 리퀘스트를 보내주시기 바랍니다.
문서 번역이나 리뷰에 참여하려면
[docs-ko@tensorflow.org](https://groups.google.com/a/tensorflow.org/forum/#!forum/docs-ko)로
메일을 보내주시기 바랍니다.

이 노트북은 영화 리뷰(review) 텍스트를 *긍정*(positive) 또는 *부정*(negative)으로 분류합니다. 이 예제는 *이진*(binary)-또는 클래스(class)가 두 개인- 분류 문제입니다. 이진 분류는 머신러닝에서 중요하고 널리 사용됩니다.

이 튜토리얼에서는 텐서플로 허브(TensorFlow Hub)와 케라스(Keras)를 사용한 기초적인 전이 학습(transfer learning) 애플리케이션을 보여줍니다.

여기에서는 [인터넷 영화 데이터베이스](https://www.imdb.com/)(Internet Movie Database)에서 수집한 50,000개의 영화 리뷰 텍스트를 담은 [IMDB 데이터셋](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb)을 사용하겠습니다. 25,000개 리뷰는 훈련용으로, 25,000개는 테스트용으로 나뉘어져 있습니다. 훈련 세트와 테스트 세트의 클래스는 *균형*이 잡혀 있습니다. 즉 긍정적인 리뷰와 부정적인 리뷰의 개수가 동일합니다.

이 노트북은 텐서플로에서 모델을 만들고 훈련하기 위한 고수준 파이썬 API인 [tf.keras](https://www.tensorflow.org/guide/keras)와 전이 학습 라이브러리이자 플랫폼인 [텐서플로 허브](https://www.tensorflow.org/hub)를 사용합니다. `tf.keras`를 사용한 고급 텍스트 분류 튜토리얼은 [MLCC 텍스트 분류 가이드](https://developers.google.com/machine-learning/guides/text-classification/)를 참고하세요.

In [3]:
import numpy as np

import tensorflow as tf

!pip install -q tensorflow-hub
!pip install -q tfds-nightly
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("버전: ", tf.__version__)
print("즉시 실행 모드: ", tf.executing_eagerly())
print("허브 버전: ", hub.__version__)
print("GPU", "사용 가능" if tf.config.experimental.list_physical_devices("GPU") else "사용 불가능")

You should consider upgrading via the '/tmpfs/src/tf_docs_env/bin/python -m pip install --upgrade pip' command.


You should consider upgrading via the '/tmpfs/src/tf_docs_env/bin/python -m pip install --upgrade pip' command.


버전:  2.3.0
즉시 실행 모드:  True
허브 버전:  0.9.0
GPU 사용 가능


## IMDB 데이터셋 다운로드

IMDB 데이터셋은 [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) 또는 [텐서플로 데이터셋](https://www.tensorflow.org/datasets)(TensorFlow datasets)에 포함되어 있습니다. 다음 코드는 IMDB 데이터셋을 컴퓨터(또는 코랩 런타임)에 다운로드합니다:

In [4]:
# 훈련 세트를 6대 4로 나눕니다.
# 결국 훈련에 15,000개 샘플, 검증에 10,000개 샘플, 테스트에 25,000개 샘플을 사용하게 됩니다.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete7YNS8F/imdb_reviews-train.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete7YNS8F/imdb_reviews-test.tfrecord


Shuffling and writing examples to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0.incomplete7YNS8F/imdb_reviews-unsupervised.tfrecord


Dataset imdb_reviews downloaded and prepared to /home/kbuilder/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


## 데이터 탐색

잠시 데이터 형태를 알아 보죠. 이 데이터셋의 샘플은 전처리된 정수 배열입니다. 이 정수는 영화 리뷰에 나오는 단어를 나타냅니다. 레이블(label)은 정수 0 또는 1입니다. 0은 부정적인 리뷰이고 1은 긍정적인 리뷰입니다.

처음 10개의 샘플을 출력해 보겠습니다.

In [5]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

처음 10개의 레이블도 출력해 보겠습니다.

In [6]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## 모델 구성

신경망은 층을 쌓아서 만듭니다. 여기에는 세 개의 중요한 구조적 결정이 필요합니다:

* 어떻게 텍스트를 표현할 것인가?
* 모델에서 얼마나 많은 층을 사용할 것인가?
* 각 층에서 얼마나 많은 *은닉 유닛*(hidden unit)을 사용할 것인가?

이 예제의 입력 데이터는 문장으로 구성됩니다. 예측할 레이블은 0 또는 1입니다.

텍스트를 표현하는 한 가지 방법은 문장을 임베딩(embedding) 벡터로 바꾸는 것입니다. 그러면 첫 번째 층으로 사전 훈련(pre-trained)된 텍스트 임베딩을 사용할 수 있습니다. 여기에는 두 가지 장점이 있습니다.

* 텍스트 전처리에 대해 신경 쓸 필요가 없습니다.
* 전이 학습의 장점을 이용합니다.
* 임베딩은 고정 크기이기 때문에 처리 과정이 단순해집니다.

이 예제는 [텐서플로 허브](https://www.tensorflow.org/hub)에 있는 **사전 훈련된 텍스트 임베딩 모델**인 [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)을 사용하겠습니다.

테스트해 볼 수 있는 사전 훈련된 모델이 세 개 더 있습니다:

* [google/tf2-preview/gnews-swivel-20dim-with-oov/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim-with-oov/1) - [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)와 동일하지만 어휘 사전(vocabulary)의 2.5%가 OOV 버킷(bucket)으로 변환되었습니다. 이는 해당 문제의 어휘 사전과 모델의 어휘 사전이 완전히 겹치지 않을 때 도움이 됩니다.
* [google/tf2-preview/nnlm-en-dim50/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1) - 더 큰 모델입니다. 차원 크기는 50이고 어휘 사전의 크기는 1백만 개 이하입니다.
* [google/tf2-preview/nnlm-en-dim128/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1) - 훨씬 더 큰 모델입니다. 차원 크기는 128이고 어휘 사전의 크기는 1백만 개 이하입니다.

먼저 문장을 임베딩시키기 위해 텐서플로 허브 모델을 사용하는 케라스 층을 만들어 보죠. 그다음 몇 개의 샘플을 입력하여 테스트해 보겠습니다. 입력 텍스트의 길이에 상관없이 임베딩의 출력 크기는 `(num_examples, embedding_dimension)`가 됩니다.

In [7]:
embedding = "https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 20), dtype=float32, numpy=
array([[ 1.765786  , -3.882232  ,  3.9134233 , -1.5557289 , -3.3362343 ,
        -1.7357955 , -1.9954445 ,  1.2989551 ,  5.081598  , -1.1041286 ,
        -2.0503852 , -0.72675157, -0.65675956,  0.24436149, -3.7208383 ,
         2.0954835 ,  2.2969332 , -2.0689783 , -2.9489717 , -1.1315987 ],
       [ 1.8804485 , -2.5852382 ,  3.4066997 ,  1.0982676 , -4.056685  ,
        -4.891284  , -2.785554  ,  1.3874227 ,  3.8476458 , -0.9256538 ,
        -1.896706  ,  1.2113281 ,  0.11474707,  0.76209456, -4.8791065 ,
         2.906149  ,  4.7087674 , -2.3652055 , -3.5015898 , -1.6390051 ],
       [ 0.71152234, -0.6353217 ,  1.7385626 , -1.1168286 , -0.5451594 ,
        -1.1808156 ,  0.09504455,  1.4653089 ,  0.66059524,  0.79308075,
        -2.2268345 ,  0.07446612, -1.4075904 , -0.70645386, -1.907037  ,
         1.4419787 ,  1.9551861 , -0.42660055, -2.8022065 ,  0.43727064]],
      dtype=float32)>

이제 전체 모델을 만들어 보겠습니다:

In [8]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 20)                400020    
_________________________________________________________________
dense (Dense)                (None, 16)                336       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 400,373
Trainable params: 400,373
Non-trainable params: 0
_________________________________________________________________


순서대로 층을 쌓아 분류기를 만듭니다:

1. 첫 번째 층은 텐서플로 허브 층입니다. 이 층은 사전 훈련된 모델을 사용하여 하나의 문장을 임베딩 벡터로 매핑합니다. 여기서 사용하는 사전 훈련된 텍스트 임베딩 모델([google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1))은 하나의 문장을 토큰(token)으로 나누고 각 토큰의 임베딩을 연결하여 반환합니다. 최종 차원은 `(num_examples, embedding_dimension)`입니다.
2. 이 고정 크기의 출력 벡터는 16개의 은닉 유닛(hidden unit)을 가진 완전 연결 층(`Dense`)으로 주입됩니다.
3. 마지막 층은 하나의 출력 노드를 가진 완전 연결 층입니다. `sigmoid` 활성화 함수를 사용하므로 확률 또는 신뢰도 수준을 표현하는 0~1 사이의 실수가 출력됩니다.

이제 모델을 컴파일합니다.

### 손실 함수와 옵티마이저

모델이 훈련하려면 손실 함수(loss function)과 옵티마이저(optimizer)가 필요합니다. 이 예제는 이진 분류 문제이고 모델이 확률을 출력하므로(출력층의 유닛이 하나이고 `sigmoid` 활성화 함수를 사용합니다), `binary_crossentropy` 손실 함수를 사용하겠습니다.

다른 손실 함수를 선택할 수 없는 것은 아닙니다. 예를 들어 `mean_squared_error`를 선택할 수 있습니다. 하지만 일반적으로 `binary_crossentropy`가 확률을 다루는데 적합합니다. 이 함수는 확률 분포 간의 거리를 측정합니다. 여기에서는 정답인 타깃 분포와 예측 분포 사이의 거리입니다.

나중에 회귀(regression) 문제(예를 들어 주택 가격을 예측하는 문제)에 대해 살펴 볼 때 평균 제곱 오차(mean squared error) 손실 함수를 어떻게 사용하는지 알아 보겠습니다.

이제 모델이 사용할 옵티마이저와 손실 함수를 설정해 보죠:

In [9]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## 모델 훈련

이 모델을 512개의 샘플로 이루어진 미니배치(mini-batch)에서 20번의 에포크(epoch) 동안 훈련합니다. `x_train`과 `y_train` 텐서에 있는 모든 샘플에 대해 20번 반복한다는 뜻입니다. 훈련하는 동안 10,000개의 검증 세트에서 모델의 손실과 정확도를 모니터링합니다:

In [10]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=20,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/20


 1/30 [>.............................] - ETA: 0s - loss: 1.1198 - accuracy: 0.4883

 3/30 [==>...........................] - ETA: 0s - loss: 1.0681 - accuracy: 0.4948

 5/30 [====>.........................] - ETA: 1s - loss: 1.0097 - accuracy: 0.4988

 7/30 [======>.......................] - ETA: 0s - loss: 0.9685 - accuracy: 0.4967

 9/30 [========>.....................] - ETA: 0s - loss: 0.9193 - accuracy: 0.5043

11/30 [==========>...................] - ETA: 0s - loss: 0.8816 - accuracy: 0.5089

13/30 [============>.................] - ETA: 0s - loss: 0.8492 - accuracy: 0.5162

15/30 [==============>...............] - ETA: 0s - loss: 0.8276 - accuracy: 0.5216

17/30 [================>.............] - ETA: 0s - loss: 0.8113 - accuracy: 0.5285

19/30 [==================>...........] - ETA: 0s - loss: 0.7976 - accuracy: 0.5329

21/30 [====================>.........] - ETA: 0s - loss: 0.7893 - accuracy: 0.5381

23/30 [======================>.......] - ETA: 0s - loss: 0.7817 - accuracy: 0.5409

25/30 [========================>.....] - ETA: 0s - loss: 0.7749 - accuracy: 0.5454

27/30 [==========================>...] - ETA: 0s - loss: 0.7664 - accuracy: 0.5500

29/30 [============================>.] - ETA: 0s - loss: 0.7601 - accuracy: 0.5535

30/30 [==============================] - 2s 64ms/step - loss: 0.7594 - accuracy: 0.5536 - val_loss: 0.6590 - val_accuracy: 0.6012


Epoch 2/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6658 - accuracy: 0.6250

 3/30 [==>...........................] - ETA: 0s - loss: 0.6617 - accuracy: 0.6094

 5/30 [====>.........................] - ETA: 0s - loss: 0.6595 - accuracy: 0.5922

 7/30 [======>.......................] - ETA: 0s - loss: 0.6591 - accuracy: 0.5837

 9/30 [========>.....................] - ETA: 0s - loss: 0.6533 - accuracy: 0.5894

11/30 [==========>...................] - ETA: 0s - loss: 0.6523 - accuracy: 0.5852

13/30 [============>.................] - ETA: 0s - loss: 0.6487 - accuracy: 0.5879

15/30 [==============>...............] - ETA: 0s - loss: 0.6463 - accuracy: 0.5896

17/30 [================>.............] - ETA: 0s - loss: 0.6459 - accuracy: 0.5915

19/30 [==================>...........] - ETA: 0s - loss: 0.6443 - accuracy: 0.5930

21/30 [====================>.........] - ETA: 0s - loss: 0.6431 - accuracy: 0.5962

23/30 [======================>.......] - ETA: 0s - loss: 0.6406 - accuracy: 0.5994

25/30 [========================>.....] - ETA: 0s - loss: 0.6397 - accuracy: 0.6030

27/30 [==========================>...] - ETA: 0s - loss: 0.6393 - accuracy: 0.6042

29/30 [============================>.] - ETA: 0s - loss: 0.6384 - accuracy: 0.6060

30/30 [==============================] - 2s 58ms/step - loss: 0.6378 - accuracy: 0.6067 - val_loss: 0.6226 - val_accuracy: 0.6354


Epoch 3/20


 1/30 [>.............................] - ETA: 0s - loss: 0.6169 - accuracy: 0.6699

 3/30 [==>...........................] - ETA: 0s - loss: 0.6137 - accuracy: 0.6504

 5/30 [====>.........................] - ETA: 0s - loss: 0.6111 - accuracy: 0.6430

 7/30 [======>.......................] - ETA: 0s - loss: 0.6097 - accuracy: 0.6437

 9/30 [========>.....................] - ETA: 0s - loss: 0.6129 - accuracy: 0.6404

11/30 [==========>...................] - ETA: 0s - loss: 0.6126 - accuracy: 0.6397

13/30 [============>.................] - ETA: 0s - loss: 0.6126 - accuracy: 0.6354

15/30 [==============>...............] - ETA: 0s - loss: 0.6125 - accuracy: 0.6368

17/30 [================>.............] - ETA: 0s - loss: 0.6079 - accuracy: 0.6403

19/30 [==================>...........] - ETA: 0s - loss: 0.6076 - accuracy: 0.6392

21/30 [====================>.........] - ETA: 0s - loss: 0.6063 - accuracy: 0.6389

23/30 [======================>.......] - ETA: 0s - loss: 0.6069 - accuracy: 0.6384

25/30 [========================>.....] - ETA: 0s - loss: 0.6060 - accuracy: 0.6413

27/30 [==========================>...] - ETA: 0s - loss: 0.6038 - accuracy: 0.6449

29/30 [============================>.] - ETA: 0s - loss: 0.6027 - accuracy: 0.6459

30/30 [==============================] - 2s 58ms/step - loss: 0.6026 - accuracy: 0.6460 - val_loss: 0.5951 - val_accuracy: 0.6601


Epoch 4/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5994 - accuracy: 0.6484

 3/30 [==>...........................] - ETA: 0s - loss: 0.5906 - accuracy: 0.6810

 5/30 [====>.........................] - ETA: 0s - loss: 0.5879 - accuracy: 0.6789

 7/30 [======>.......................] - ETA: 0s - loss: 0.5903 - accuracy: 0.6699

 9/30 [========>.....................] - ETA: 0s - loss: 0.5880 - accuracy: 0.6708

11/30 [==========>...................] - ETA: 0s - loss: 0.5857 - accuracy: 0.6712

13/30 [============>.................] - ETA: 0s - loss: 0.5839 - accuracy: 0.6716

15/30 [==============>...............] - ETA: 0s - loss: 0.5833 - accuracy: 0.6697

17/30 [================>.............] - ETA: 0s - loss: 0.5804 - accuracy: 0.6704

19/30 [==================>...........] - ETA: 0s - loss: 0.5806 - accuracy: 0.6708

21/30 [====================>.........] - ETA: 0s - loss: 0.5790 - accuracy: 0.6704

23/30 [======================>.......] - ETA: 0s - loss: 0.5771 - accuracy: 0.6731

25/30 [========================>.....] - ETA: 0s - loss: 0.5749 - accuracy: 0.6752

27/30 [==========================>...] - ETA: 0s - loss: 0.5737 - accuracy: 0.6769

29/30 [============================>.] - ETA: 0s - loss: 0.5719 - accuracy: 0.6783

30/30 [==============================] - 2s 58ms/step - loss: 0.5718 - accuracy: 0.6790 - val_loss: 0.5683 - val_accuracy: 0.6911


Epoch 5/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5539 - accuracy: 0.6797

 3/30 [==>...........................] - ETA: 0s - loss: 0.5644 - accuracy: 0.6953

 5/30 [====>.........................] - ETA: 0s - loss: 0.5628 - accuracy: 0.6961

 7/30 [======>.......................] - ETA: 0s - loss: 0.5576 - accuracy: 0.6942

 9/30 [========>.....................] - ETA: 0s - loss: 0.5570 - accuracy: 0.6951

11/30 [==========>...................] - ETA: 0s - loss: 0.5564 - accuracy: 0.6914

13/30 [============>.................] - ETA: 0s - loss: 0.5565 - accuracy: 0.6892

15/30 [==============>...............] - ETA: 0s - loss: 0.5534 - accuracy: 0.6940

17/30 [================>.............] - ETA: 0s - loss: 0.5523 - accuracy: 0.6973

19/30 [==================>...........] - ETA: 0s - loss: 0.5503 - accuracy: 0.6993

21/30 [====================>.........] - ETA: 0s - loss: 0.5479 - accuracy: 0.7021

23/30 [======================>.......] - ETA: 0s - loss: 0.5472 - accuracy: 0.7034

25/30 [========================>.....] - ETA: 0s - loss: 0.5449 - accuracy: 0.7064

27/30 [==========================>...] - ETA: 0s - loss: 0.5423 - accuracy: 0.7098

29/30 [============================>.] - ETA: 0s - loss: 0.5400 - accuracy: 0.7126

30/30 [==============================] - 2s 58ms/step - loss: 0.5400 - accuracy: 0.7126 - val_loss: 0.5402 - val_accuracy: 0.6958


Epoch 6/20


 1/30 [>.............................] - ETA: 0s - loss: 0.5137 - accuracy: 0.7207

 3/30 [==>...........................] - ETA: 0s - loss: 0.5124 - accuracy: 0.7188

 5/30 [====>.........................] - ETA: 0s - loss: 0.5116 - accuracy: 0.7199

 7/30 [======>.......................] - ETA: 0s - loss: 0.5121 - accuracy: 0.7157

 9/30 [========>.....................] - ETA: 0s - loss: 0.5121 - accuracy: 0.7214

11/30 [==========>...................] - ETA: 0s - loss: 0.5129 - accuracy: 0.7232

13/30 [============>.................] - ETA: 0s - loss: 0.5114 - accuracy: 0.7299

15/30 [==============>...............] - ETA: 0s - loss: 0.5097 - accuracy: 0.7332

17/30 [================>.............] - ETA: 0s - loss: 0.5101 - accuracy: 0.7308

19/30 [==================>...........] - ETA: 0s - loss: 0.5108 - accuracy: 0.7301

21/30 [====================>.........] - ETA: 0s - loss: 0.5104 - accuracy: 0.7305

23/30 [======================>.......] - ETA: 0s - loss: 0.5093 - accuracy: 0.7321

25/30 [========================>.....] - ETA: 0s - loss: 0.5103 - accuracy: 0.7320

27/30 [==========================>...] - ETA: 0s - loss: 0.5081 - accuracy: 0.7337

29/30 [============================>.] - ETA: 0s - loss: 0.5070 - accuracy: 0.7345

30/30 [==============================] - 2s 58ms/step - loss: 0.5070 - accuracy: 0.7347 - val_loss: 0.5085 - val_accuracy: 0.7300


Epoch 7/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4747 - accuracy: 0.7520

 3/30 [==>...........................] - ETA: 0s - loss: 0.4888 - accuracy: 0.7409

 5/30 [====>.........................] - ETA: 0s - loss: 0.4838 - accuracy: 0.7457

 7/30 [======>.......................] - ETA: 0s - loss: 0.4815 - accuracy: 0.7511

 9/30 [========>.....................] - ETA: 0s - loss: 0.4822 - accuracy: 0.7552

11/30 [==========>...................] - ETA: 0s - loss: 0.4797 - accuracy: 0.7589

13/30 [============>.................] - ETA: 0s - loss: 0.4788 - accuracy: 0.7601

15/30 [==============>...............] - ETA: 0s - loss: 0.4762 - accuracy: 0.7617

17/30 [================>.............] - ETA: 0s - loss: 0.4771 - accuracy: 0.7593

19/30 [==================>...........] - ETA: 0s - loss: 0.4756 - accuracy: 0.7617

21/30 [====================>.........] - ETA: 0s - loss: 0.4743 - accuracy: 0.7629

23/30 [======================>.......] - ETA: 0s - loss: 0.4742 - accuracy: 0.7645

25/30 [========================>.....] - ETA: 0s - loss: 0.4719 - accuracy: 0.7655

27/30 [==========================>...] - ETA: 0s - loss: 0.4708 - accuracy: 0.7663

29/30 [============================>.] - ETA: 0s - loss: 0.4701 - accuracy: 0.7664

30/30 [==============================] - 2s 59ms/step - loss: 0.4698 - accuracy: 0.7667 - val_loss: 0.4756 - val_accuracy: 0.7657


Epoch 8/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4518 - accuracy: 0.7773

 3/30 [==>...........................] - ETA: 0s - loss: 0.4503 - accuracy: 0.7845

 5/30 [====>.........................] - ETA: 0s - loss: 0.4440 - accuracy: 0.7922

 7/30 [======>.......................] - ETA: 0s - loss: 0.4422 - accuracy: 0.7868

 9/30 [========>.....................] - ETA: 0s - loss: 0.4428 - accuracy: 0.7871

11/30 [==========>...................] - ETA: 0s - loss: 0.4416 - accuracy: 0.7866

13/30 [============>.................] - ETA: 0s - loss: 0.4430 - accuracy: 0.7843

15/30 [==============>...............] - ETA: 0s - loss: 0.4369 - accuracy: 0.7892

17/30 [================>.............] - ETA: 0s - loss: 0.4379 - accuracy: 0.7876

19/30 [==================>...........] - ETA: 0s - loss: 0.4367 - accuracy: 0.7902

21/30 [====================>.........] - ETA: 0s - loss: 0.4354 - accuracy: 0.7922

23/30 [======================>.......] - ETA: 0s - loss: 0.4349 - accuracy: 0.7938

25/30 [========================>.....] - ETA: 0s - loss: 0.4344 - accuracy: 0.7946

27/30 [==========================>...] - ETA: 0s - loss: 0.4327 - accuracy: 0.7964

29/30 [============================>.] - ETA: 0s - loss: 0.4318 - accuracy: 0.7961

30/30 [==============================] - 2s 58ms/step - loss: 0.4314 - accuracy: 0.7965 - val_loss: 0.4417 - val_accuracy: 0.7861


Epoch 9/20


 1/30 [>.............................] - ETA: 0s - loss: 0.4009 - accuracy: 0.8203

 3/30 [==>...........................] - ETA: 0s - loss: 0.4121 - accuracy: 0.8034

 5/30 [====>.........................] - ETA: 0s - loss: 0.4057 - accuracy: 0.8047

 7/30 [======>.......................] - ETA: 0s - loss: 0.4099 - accuracy: 0.8050

 9/30 [========>.....................] - ETA: 0s - loss: 0.4063 - accuracy: 0.8084

11/30 [==========>...................] - ETA: 0s - loss: 0.4009 - accuracy: 0.8125

13/30 [============>.................] - ETA: 0s - loss: 0.4027 - accuracy: 0.8139

15/30 [==============>...............] - ETA: 0s - loss: 0.3990 - accuracy: 0.8193

17/30 [================>.............] - ETA: 0s - loss: 0.3980 - accuracy: 0.8193

19/30 [==================>...........] - ETA: 0s - loss: 0.3980 - accuracy: 0.8183

21/30 [====================>.........] - ETA: 0s - loss: 0.3998 - accuracy: 0.8165

23/30 [======================>.......] - ETA: 0s - loss: 0.3972 - accuracy: 0.8172

25/30 [========================>.....] - ETA: 0s - loss: 0.3958 - accuracy: 0.8180

27/30 [==========================>...] - ETA: 0s - loss: 0.3934 - accuracy: 0.8191

29/30 [============================>.] - ETA: 0s - loss: 0.3923 - accuracy: 0.8198

30/30 [==============================] - 2s 58ms/step - loss: 0.3920 - accuracy: 0.8199 - val_loss: 0.4113 - val_accuracy: 0.8160


Epoch 10/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3804 - accuracy: 0.8242

 3/30 [==>...........................] - ETA: 0s - loss: 0.3759 - accuracy: 0.8398

 5/30 [====>.........................] - ETA: 0s - loss: 0.3737 - accuracy: 0.8418

 7/30 [======>.......................] - ETA: 0s - loss: 0.3757 - accuracy: 0.8401

 9/30 [========>.....................] - ETA: 0s - loss: 0.3729 - accuracy: 0.8409

11/30 [==========>...................] - ETA: 0s - loss: 0.3685 - accuracy: 0.8425

13/30 [============>.................] - ETA: 0s - loss: 0.3622 - accuracy: 0.8451

15/30 [==============>...............] - ETA: 0s - loss: 0.3619 - accuracy: 0.8430

17/30 [================>.............] - ETA: 0s - loss: 0.3598 - accuracy: 0.8442

19/30 [==================>...........] - ETA: 0s - loss: 0.3603 - accuracy: 0.8434

21/30 [====================>.........] - ETA: 0s - loss: 0.3609 - accuracy: 0.8418

23/30 [======================>.......] - ETA: 0s - loss: 0.3593 - accuracy: 0.8432

25/30 [========================>.....] - ETA: 0s - loss: 0.3566 - accuracy: 0.8446

27/30 [==========================>...] - ETA: 0s - loss: 0.3568 - accuracy: 0.8439

29/30 [============================>.] - ETA: 0s - loss: 0.3556 - accuracy: 0.8437

30/30 [==============================] - 2s 57ms/step - loss: 0.3552 - accuracy: 0.8441 - val_loss: 0.3840 - val_accuracy: 0.8292


Epoch 11/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3493 - accuracy: 0.8477

 3/30 [==>...........................] - ETA: 0s - loss: 0.3325 - accuracy: 0.8594

 5/30 [====>.........................] - ETA: 0s - loss: 0.3351 - accuracy: 0.8531

 7/30 [======>.......................] - ETA: 0s - loss: 0.3325 - accuracy: 0.8499

 9/30 [========>.....................] - ETA: 0s - loss: 0.3302 - accuracy: 0.8548

11/30 [==========>...................] - ETA: 0s - loss: 0.3301 - accuracy: 0.8544

13/30 [============>.................] - ETA: 0s - loss: 0.3284 - accuracy: 0.8549

15/30 [==============>...............] - ETA: 0s - loss: 0.3243 - accuracy: 0.8581

17/30 [================>.............] - ETA: 0s - loss: 0.3232 - accuracy: 0.8591

19/30 [==================>...........] - ETA: 0s - loss: 0.3216 - accuracy: 0.8616

21/30 [====================>.........] - ETA: 0s - loss: 0.3243 - accuracy: 0.8605

23/30 [======================>.......] - ETA: 0s - loss: 0.3237 - accuracy: 0.8608

25/30 [========================>.....] - ETA: 0s - loss: 0.3233 - accuracy: 0.8616

27/30 [==========================>...] - ETA: 0s - loss: 0.3226 - accuracy: 0.8620

29/30 [============================>.] - ETA: 0s - loss: 0.3224 - accuracy: 0.8622

30/30 [==============================] - 2s 58ms/step - loss: 0.3229 - accuracy: 0.8619 - val_loss: 0.3631 - val_accuracy: 0.8361


Epoch 12/20


 1/30 [>.............................] - ETA: 0s - loss: 0.3158 - accuracy: 0.8652

 3/30 [==>...........................] - ETA: 0s - loss: 0.3072 - accuracy: 0.8802

 5/30 [====>.........................] - ETA: 0s - loss: 0.3067 - accuracy: 0.8695

 7/30 [======>.......................] - ETA: 0s - loss: 0.3111 - accuracy: 0.8658

 9/30 [========>.....................] - ETA: 0s - loss: 0.3077 - accuracy: 0.8674

11/30 [==========>...................] - ETA: 0s - loss: 0.3054 - accuracy: 0.8704

13/30 [============>.................] - ETA: 0s - loss: 0.3047 - accuracy: 0.8732

15/30 [==============>...............] - ETA: 0s - loss: 0.3021 - accuracy: 0.8755

17/30 [================>.............] - ETA: 0s - loss: 0.3007 - accuracy: 0.8748

19/30 [==================>...........] - ETA: 0s - loss: 0.2994 - accuracy: 0.8753

21/30 [====================>.........] - ETA: 0s - loss: 0.2986 - accuracy: 0.8744

23/30 [======================>.......] - ETA: 0s - loss: 0.2973 - accuracy: 0.8757

25/30 [========================>.....] - ETA: 0s - loss: 0.2958 - accuracy: 0.8763

27/30 [==========================>...] - ETA: 0s - loss: 0.2968 - accuracy: 0.8758

29/30 [============================>.] - ETA: 0s - loss: 0.2966 - accuracy: 0.8765

30/30 [==============================] - 2s 58ms/step - loss: 0.2961 - accuracy: 0.8765 - val_loss: 0.3474 - val_accuracy: 0.8475


Epoch 13/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2907 - accuracy: 0.8828

 3/30 [==>...........................] - ETA: 0s - loss: 0.2962 - accuracy: 0.8822

 5/30 [====>.........................] - ETA: 0s - loss: 0.2839 - accuracy: 0.8891

 7/30 [======>.......................] - ETA: 0s - loss: 0.2829 - accuracy: 0.8881

 9/30 [========>.....................] - ETA: 0s - loss: 0.2777 - accuracy: 0.8919

11/30 [==========>...................] - ETA: 0s - loss: 0.2799 - accuracy: 0.8885

13/30 [============>.................] - ETA: 0s - loss: 0.2779 - accuracy: 0.8887

15/30 [==============>...............] - ETA: 0s - loss: 0.2745 - accuracy: 0.8901

17/30 [================>.............] - ETA: 0s - loss: 0.2741 - accuracy: 0.8901

19/30 [==================>...........] - ETA: 0s - loss: 0.2751 - accuracy: 0.8893

21/30 [====================>.........] - ETA: 0s - loss: 0.2748 - accuracy: 0.8893

23/30 [======================>.......] - ETA: 0s - loss: 0.2738 - accuracy: 0.8894

25/30 [========================>.....] - ETA: 0s - loss: 0.2726 - accuracy: 0.8891

27/30 [==========================>...] - ETA: 0s - loss: 0.2731 - accuracy: 0.8893

29/30 [============================>.] - ETA: 0s - loss: 0.2728 - accuracy: 0.8900

30/30 [==============================] - 2s 58ms/step - loss: 0.2725 - accuracy: 0.8903 - val_loss: 0.3350 - val_accuracy: 0.8523


Epoch 14/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2702 - accuracy: 0.8926

 3/30 [==>...........................] - ETA: 0s - loss: 0.2559 - accuracy: 0.8932

 5/30 [====>.........................] - ETA: 0s - loss: 0.2569 - accuracy: 0.8895

 7/30 [======>.......................] - ETA: 0s - loss: 0.2543 - accuracy: 0.8909

 9/30 [========>.....................] - ETA: 0s - loss: 0.2536 - accuracy: 0.8928

11/30 [==========>...................] - ETA: 0s - loss: 0.2544 - accuracy: 0.8951

13/30 [============>.................] - ETA: 0s - loss: 0.2528 - accuracy: 0.8978

15/30 [==============>...............] - ETA: 0s - loss: 0.2540 - accuracy: 0.8973

17/30 [================>.............] - ETA: 0s - loss: 0.2532 - accuracy: 0.8967

19/30 [==================>...........] - ETA: 0s - loss: 0.2541 - accuracy: 0.8951

21/30 [====================>.........] - ETA: 0s - loss: 0.2539 - accuracy: 0.8947

23/30 [======================>.......] - ETA: 0s - loss: 0.2535 - accuracy: 0.8959

25/30 [========================>.....] - ETA: 0s - loss: 0.2525 - accuracy: 0.8967

27/30 [==========================>...] - ETA: 0s - loss: 0.2525 - accuracy: 0.8971

29/30 [============================>.] - ETA: 0s - loss: 0.2525 - accuracy: 0.8969

30/30 [==============================] - 2s 58ms/step - loss: 0.2524 - accuracy: 0.8967 - val_loss: 0.3280 - val_accuracy: 0.8519


Epoch 15/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2481 - accuracy: 0.8926

 3/30 [==>...........................] - ETA: 0s - loss: 0.2410 - accuracy: 0.8952

 5/30 [====>.........................] - ETA: 0s - loss: 0.2351 - accuracy: 0.9004

 7/30 [======>.......................] - ETA: 0s - loss: 0.2348 - accuracy: 0.9062

 9/30 [========>.....................] - ETA: 0s - loss: 0.2381 - accuracy: 0.9045

11/30 [==========>...................] - ETA: 0s - loss: 0.2347 - accuracy: 0.9057

13/30 [============>.................] - ETA: 0s - loss: 0.2345 - accuracy: 0.9052

15/30 [==============>...............] - ETA: 0s - loss: 0.2339 - accuracy: 0.9056

17/30 [================>.............] - ETA: 0s - loss: 0.2337 - accuracy: 0.9064

19/30 [==================>...........] - ETA: 0s - loss: 0.2343 - accuracy: 0.9062

21/30 [====================>.........] - ETA: 0s - loss: 0.2353 - accuracy: 0.9057

23/30 [======================>.......] - ETA: 0s - loss: 0.2355 - accuracy: 0.9056

25/30 [========================>.....] - ETA: 0s - loss: 0.2349 - accuracy: 0.9069

27/30 [==========================>...] - ETA: 0s - loss: 0.2339 - accuracy: 0.9070

29/30 [============================>.] - ETA: 0s - loss: 0.2350 - accuracy: 0.9062

30/30 [==============================] - 2s 58ms/step - loss: 0.2347 - accuracy: 0.9065 - val_loss: 0.3191 - val_accuracy: 0.8591


Epoch 16/20


 1/30 [>.............................] - ETA: 0s - loss: 0.2365 - accuracy: 0.9043

 3/30 [==>...........................] - ETA: 0s - loss: 0.2254 - accuracy: 0.9082

 5/30 [====>.........................] - ETA: 0s - loss: 0.2169 - accuracy: 0.9172

 7/30 [======>.......................] - ETA: 0s - loss: 0.2218 - accuracy: 0.9138

 9/30 [========>.....................] - ETA: 0s - loss: 0.2250 - accuracy: 0.9141

11/30 [==========>...................] - ETA: 0s - loss: 0.2201 - accuracy: 0.9165

13/30 [============>.................] - ETA: 0s - loss: 0.2185 - accuracy: 0.9174

15/30 [==============>...............] - ETA: 0s - loss: 0.2196 - accuracy: 0.9159

17/30 [================>.............] - ETA: 0s - loss: 0.2217 - accuracy: 0.9142

19/30 [==================>...........] - ETA: 0s - loss: 0.2203 - accuracy: 0.9148

21/30 [====================>.........] - ETA: 0s - loss: 0.2195 - accuracy: 0.9156

23/30 [======================>.......] - ETA: 0s - loss: 0.2188 - accuracy: 0.9168

25/30 [========================>.....] - ETA: 0s - loss: 0.2188 - accuracy: 0.9164

27/30 [==========================>...] - ETA: 0s - loss: 0.2181 - accuracy: 0.9162

29/30 [============================>.] - ETA: 0s - loss: 0.2178 - accuracy: 0.9159

30/30 [==============================] - 2s 58ms/step - loss: 0.2184 - accuracy: 0.9157 - val_loss: 0.3136 - val_accuracy: 0.8639


Epoch 17/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1720 - accuracy: 0.9473

 3/30 [==>...........................] - ETA: 0s - loss: 0.2030 - accuracy: 0.9284

 5/30 [====>.........................] - ETA: 0s - loss: 0.2101 - accuracy: 0.9156

 7/30 [======>.......................] - ETA: 0s - loss: 0.2088 - accuracy: 0.9185

 9/30 [========>.....................] - ETA: 0s - loss: 0.2108 - accuracy: 0.9180

11/30 [==========>...................] - ETA: 0s - loss: 0.2085 - accuracy: 0.9199

13/30 [============>.................] - ETA: 0s - loss: 0.2104 - accuracy: 0.9190

15/30 [==============>...............] - ETA: 0s - loss: 0.2083 - accuracy: 0.9208

17/30 [================>.............] - ETA: 0s - loss: 0.2082 - accuracy: 0.9207

19/30 [==================>...........] - ETA: 0s - loss: 0.2062 - accuracy: 0.9200

21/30 [====================>.........] - ETA: 0s - loss: 0.2051 - accuracy: 0.9207

23/30 [======================>.......] - ETA: 0s - loss: 0.2036 - accuracy: 0.9220

25/30 [========================>.....] - ETA: 0s - loss: 0.2030 - accuracy: 0.9228

27/30 [==========================>...] - ETA: 0s - loss: 0.2033 - accuracy: 0.9223

29/30 [============================>.] - ETA: 0s - loss: 0.2040 - accuracy: 0.9220

30/30 [==============================] - 2s 57ms/step - loss: 0.2044 - accuracy: 0.9220 - val_loss: 0.3098 - val_accuracy: 0.8657


Epoch 18/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1966 - accuracy: 0.9238

 3/30 [==>...........................] - ETA: 0s - loss: 0.1806 - accuracy: 0.9303

 5/30 [====>.........................] - ETA: 0s - loss: 0.1930 - accuracy: 0.9223

 7/30 [======>.......................] - ETA: 0s - loss: 0.1928 - accuracy: 0.9235

 9/30 [========>.....................] - ETA: 0s - loss: 0.1946 - accuracy: 0.9232

11/30 [==========>...................] - ETA: 0s - loss: 0.1938 - accuracy: 0.9254

13/30 [============>.................] - ETA: 0s - loss: 0.1951 - accuracy: 0.9256

15/30 [==============>...............] - ETA: 0s - loss: 0.1972 - accuracy: 0.9255

17/30 [================>.............] - ETA: 0s - loss: 0.1960 - accuracy: 0.9264

19/30 [==================>...........] - ETA: 0s - loss: 0.1946 - accuracy: 0.9262

21/30 [====================>.........] - ETA: 0s - loss: 0.1924 - accuracy: 0.9262

23/30 [======================>.......] - ETA: 0s - loss: 0.1924 - accuracy: 0.9258

25/30 [========================>.....] - ETA: 0s - loss: 0.1927 - accuracy: 0.9260

27/30 [==========================>...] - ETA: 0s - loss: 0.1931 - accuracy: 0.9260

29/30 [============================>.] - ETA: 0s - loss: 0.1925 - accuracy: 0.9265

30/30 [==============================] - 2s 58ms/step - loss: 0.1926 - accuracy: 0.9266 - val_loss: 0.3099 - val_accuracy: 0.8716


Epoch 19/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1748 - accuracy: 0.9395

 3/30 [==>...........................] - ETA: 0s - loss: 0.1804 - accuracy: 0.9375

 5/30 [====>.........................] - ETA: 0s - loss: 0.1733 - accuracy: 0.9395

 7/30 [======>.......................] - ETA: 0s - loss: 0.1786 - accuracy: 0.9339

 9/30 [========>.....................] - ETA: 0s - loss: 0.1788 - accuracy: 0.9329

11/30 [==========>...................] - ETA: 0s - loss: 0.1768 - accuracy: 0.9334

13/30 [============>.................] - ETA: 0s - loss: 0.1774 - accuracy: 0.9349

15/30 [==============>...............] - ETA: 0s - loss: 0.1780 - accuracy: 0.9342

17/30 [================>.............] - ETA: 0s - loss: 0.1789 - accuracy: 0.9337

19/30 [==================>...........] - ETA: 0s - loss: 0.1780 - accuracy: 0.9342

21/30 [====================>.........] - ETA: 0s - loss: 0.1776 - accuracy: 0.9333

23/30 [======================>.......] - ETA: 0s - loss: 0.1793 - accuracy: 0.9318

25/30 [========================>.....] - ETA: 0s - loss: 0.1794 - accuracy: 0.9323

27/30 [==========================>...] - ETA: 0s - loss: 0.1784 - accuracy: 0.9335

29/30 [============================>.] - ETA: 0s - loss: 0.1797 - accuracy: 0.9332

30/30 [==============================] - 2s 58ms/step - loss: 0.1794 - accuracy: 0.9333 - val_loss: 0.3061 - val_accuracy: 0.8672


Epoch 20/20


 1/30 [>.............................] - ETA: 0s - loss: 0.1524 - accuracy: 0.9414

 3/30 [==>...........................] - ETA: 0s - loss: 0.1572 - accuracy: 0.9382

 5/30 [====>.........................] - ETA: 0s - loss: 0.1504 - accuracy: 0.9457

 7/30 [======>.......................] - ETA: 0s - loss: 0.1570 - accuracy: 0.9434

 9/30 [========>.....................] - ETA: 0s - loss: 0.1610 - accuracy: 0.9427

11/30 [==========>...................] - ETA: 0s - loss: 0.1639 - accuracy: 0.9409

13/30 [============>.................] - ETA: 0s - loss: 0.1618 - accuracy: 0.9401

15/30 [==============>...............] - ETA: 0s - loss: 0.1658 - accuracy: 0.9384

17/30 [================>.............] - ETA: 0s - loss: 0.1654 - accuracy: 0.9386

19/30 [==================>...........] - ETA: 0s - loss: 0.1671 - accuracy: 0.9369

21/30 [====================>.........] - ETA: 0s - loss: 0.1685 - accuracy: 0.9366

23/30 [======================>.......] - ETA: 0s - loss: 0.1701 - accuracy: 0.9367

25/30 [========================>.....] - ETA: 0s - loss: 0.1694 - accuracy: 0.9377

27/30 [==========================>...] - ETA: 0s - loss: 0.1703 - accuracy: 0.9372

29/30 [============================>.] - ETA: 0s - loss: 0.1691 - accuracy: 0.9376

30/30 [==============================] - 2s 58ms/step - loss: 0.1687 - accuracy: 0.9379 - val_loss: 0.3065 - val_accuracy: 0.8685


## 모델 평가

모델의 성능을 확인해 보죠. 두 개의 값이 반환됩니다. 손실(오차를 나타내는 숫자이므로 낮을수록 좋습니다)과 정확도입니다.

In [11]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3206 - accuracy: 0.8550


loss: 0.321
accuracy: 0.855


이 예제는 매우 단순한 방식을 사용하므로 87% 정도의 정확도를 달성했습니다. 고급 방법을 사용한 모델은 95%에 가까운 정확도를 얻습니다.

## 더 읽을거리

문자열 입력을 다루는 조금 더 일반적인 방법과 훈련 과정에서 정확도나 손실 변화에 대한 자세한 분석은 [여기](https://www.tensorflow.org/tutorials/keras/basic_text_classification)를 참고하세요.